### ISSUE TITLE
Accessibility issues in React Aria library

### ISSUE THREAD

**Issue by guyhickling** [adobe/react-spectrum#5630]
### Provide your feedback here.

I have just taken a quick look at your React Aria release. The web dev industry is badly in need of a robust, fully accessible and WCAG-compliant React component library and, when I saw your offering, I hoped that you had done just that. Sadly, it does not look to me that you are achieving it yet, though with a little work you can make it so. 

As an accessibility auditor and consultant, I am finding numerous accessibility issues in it. None that are too serious as to make things inoperable to disabled users (at least I've found none so far), but enough to stop me recommending this library to anyone. And ones which, accessibility audits are done of websites created by your customers using this tool, will cause accessibility consultants doing the audits to fail or at least call out various components. (That will require your customers to have to amend components themselves in situ, or enter in to tedious correspondence with your team to get you to resolve the issues. I.e. time wasted for all concerned.) And it will cause loss of confidence in your library.

For instance:
1) Your toggle Button is marked up in the HTML as a link, not a <button> element, a very common mistake made by developers and one which many accessibility auditors, including myself, will raise as an issue. This will cause screen readers to announce them as links instead of buttons, which will be confusing to many blind people.
2) Your table component uses ARIA roles, even though it already uses the correct HTML elements of table, th  and td. The first Rule of ARIA is, don’t use ARIA when you can use valid HTML, so you should follow that and remove this ARIA. It will alter the announcements that some screen readers make.
3) Your Calendar component has given the month/year text, shown to sighted users on the top row of the calendar, an aria-hidden=”true” attribute. This will hide it from screen reader users completely. Why would you want to hide displayed text from blind people – they want to hear the same information that sighted users are given. I appreciate the next line of HTML, a div with the role “application”, has that information in it; nevertheless there is no need to hide the displayed text. The “application” in this case is a calendar, not a particular month, so the ARIA label should say “Calendar” to tell blind people exactly what component they have found – a calendar with many months and years, not just a widget for a single month. 
4) Also the next month and previous month arrow buttons in the Calendar have ARIA labels of just Next and Previous. Next what? – sighted users can see and immediately deduce what they are referring to, from their location on screen etc, but blind people need better context. Re-label these buttons “Next month” and “Previous month”.

These are just a few points that occur to me from just a very quick look, and I am sure there are others waiting to be found. I do suggest you tighten up on things like these to avoid your customers having your components failed or called out in professional accessibility audits. That will save them unwanted work, and avoid repeated queries to your team from your customers.

### 🔦 Context

This will affect disabled people trying to use websites created using your library. And I would like to see a robust, fully accessible and WCAG-compliant React component library, one that I can recommend to people in the accessibility industry, and to React developers generally.

### Version

any

### What browsers are you seeing the problem on?

Firefox

### What operating system are you using?

MS Windows 10

**Comment by devongovett** [comment]
Thanks for your feedback. I would recommend testing with real assistive technologies rather than simply looking at the DOM. Many of the things you cited were specifically done to work around issues found when testing on various devices or to improve the user experience with AT.

1. I'm not seeing the toggle button link issue you mention. It's a button element with aria-pressed, following the [ARIA pattern](https://www.w3.org/WAI/ARIA/apg/patterns/button/). I did a quick test with VoiceOver and it announced as a toggle button.

2. The Table component is an interactive grid, not a simple table. It supports things like sorting, keyboard navigation, row selection, etc. This requires use of the [grid role](https://www.w3.org/WAI/ARIA/apg/patterns/grid/) rather than a plain [table](https://www.w3.org/WAI/ARIA/apg/patterns/table/) which is defined as non-interactive.

3. The visible month text is aria-hidden because there is an additional visually hidden duplicate at the start of the calendar. This ensures that it is the first element in the calendar even if the next/previous buttons are visually first. That helps mobile screen readers such as VoiceOver on iOS which will not navigate or announce the aria label for the whole grid. This way users get the context immediately when swiping to the calendar. Our [blog post](https://react-spectrum.adobe.com/blog/date-and-time-pickers-for-all.html) includes more details about this including a video of the mobile screen reader experience.

4. Agreed this could be improved. The exact amount of dates that the next/previous buttons advance may also differ so it isn't always by month.

We certainly take testing pretty seriously and the components have also been included in audits as part of our products, but please let us know if you find other issues when testing with real AT!
